### ISSUE TITLE
Underscore _.where function return only one item

### ISSUE THREAD

**Issue by LionelGuerin** [meteor/meteor#13049]
Hi,

I'm using and reproducing this with meteor on Mac (version 14.2.1) and ubuntu (version 20.04.6)

Since Meteor 2.15, with the upgrade of underscore from 1.0.13 to 1.6.1 .

The return of the function _.where(array, {criteriaToFind}) seems to always be the first element to correspond to the search.

In previous meteor versions, the function _.where used to return an array of every items find.

I made a repository to reproduce the problem.

https://github.com/LionelGuerin/meteor_underscore_where

It's composed of three folders to tests with meteor 2.15 , meteor 2.14 and with vanilla underscore NPM.

### For meteor 2.15

```
git clone git@github.com:LionelGuerin/meteor_underscore_where.git

cd underscore_where/meteor_underscore

npm i

meteor
```

Go to your browser then click on "Click Me"
You should see the following **server** log

```
underscoreTest { randomid: '0DVVuahGdr', tofind: true }
```

### For meteor 2.14

```
git clone git@github.com:LionelGuerin/meteor_underscore_where.git

cd underscore_where/2-14_meteor_underscore

npm i

meteor
```

Go to your browser then click on "Click Me"
You should see the following **server** log

```
underscoreTest [
  { randomid: '0DVVuahGdr', tofind: true },
  { randomid: '6DAPnPUk7k', tofind: true }
]
```

### For vanilla underscore NPM

```
git clone git@github.com:LionelGuerin/meteor_underscore_where.git

cd underscore_where/2-14_meteor_underscore

npm i

node underscore_vanilla.js
```

You should see the following log

```
underscoreTest [
  { randomid: '0DVVuahGdr', tofind: true },
  { randomid: '6DAPnPUk7k', tofind: true }
]
```

____

I have seen another underscore issue with _.intersection, but it seems to be a different problem.
If you need any more information, i will be happy to help ;)

**Comment by allenfuller** [comment]
This hit me too -- lots of places where I was using `_.where` to filter an array, now they all return an object. Happy to help with a bug fix.

**Comment by harveysanders** [comment]
Yep, we're seeing this bug pop up in dependants of `meteor/underscore`. One example is `aldeed/autoform` `v6.3.0`.
https://github.com/Meteor-Community-Packages/meteor-autoform/blob/d8f8b6a6e36079c054976ca8bda7bec97f548b4f/autoform-api.js#L504-L505

An array is expected here, but an object is returned from `_.where()`, so we're left with
```js
_.where(/* ... */)[0]

{ /* ... * /}[0]  // <- undefined
```
Yes, I know `aldeed/autoform@6.3.0` is outdated 😅 and we're working on updating it now, but it sounds like there are other affected dependants :). 
Would it be possible to revert the `_.where` change? We could help with a PR if you like.

Here's the `where` changeset:
https://github.com/meteor/meteor/commit/47949d971c68f40ea4f52f215438fd2564f26d87#diff-1cb90728ba18568f3c72e6685a2efa7a5926eb93b30991df6a3df86d2ed5da0cL266-L273


Thanks for all the hard work!

**Comment by harveysanders** [comment]
Just in case anyone needs a short-term workaround - Meteor allows for override packages, even core, by placing the package source code in the `packages` folder in your application's root. 
```
[your_application_root]/packages/[package-name]

ex:
[your_application_root]/packages/underscore/
```
https://guide.meteor.com/writing-atmosphere-packages#local-packages

**Comment by nachocodoner** [comment]
@StorytellerCZ: Can you provide more context about this change?

It replaced `_.where` with _.find, making it return a single element instead of a filtered array. Here's the [commit](https://github.com/meteor/meteor/commit/47949d971c68f40ea4f52f215438fd2564f26d87#diff-1cb90728ba18568f3c72e6685a2efa7a5926eb93b30991df6a3df86d2ed5da0cR268). Would it be better to revert it and use `_.where` instead?

**Comment by evolross** [comment]
Dang... total annihilation of `_.where` in the official Meteor `underscore` package, which in my Meteor 2.16 app alone is depended on by these packages:

ddp-server@2.7.1
routepolicy@1.1.1
alanning:roles@1.3.0
cultofcoders:redis-oplog@2.2.1
edgee:slingshot@0.7.2

**Comment by evolross** [comment]
`_.findWhere` is also broken, likely since it uses `_.where`. Oddly, it returns an array of values instead of just one, unlike how `_.where` is returning one value instead of an array.

**Expected Results**

Looks through the list and returns the first value that all of the key-value pairs listed in properties.

For example:
```
_.findWhere(movies, {title: 'Fight Club'});

{
   title: 'Fight Club',
   year: '1999',
   rating: 5
}
```

**Current Results**

Returns the first matching value as the first element in an array.

```
_.findWhere(movies, {title: 'Fight Club'});

[
   {
      title: 'Fight Club',
      year: '1999',
      rating: 5
   }
]
```

**Comment by nachocodoner** [comment]
`meteor/underscore` is primarily a utility package for apps and other packages, not the core. This is why tests on Meteor doesn't trigger failures as part of the test checks.

I understand the concerns with these failures. I considered suggesting fixes in the next underscore upgrade (https://github.com/meteor/meteor/pull/12994) by @StorytellerCZ, but the amount of changes there makes it risky. The package lacks test coverage, making it prone to breaking dependencies, especially since it serves as a utility for many others.

One option is to create a new patch version and publish `meteor/underscore`, either reverting the breaking changes or applying your suggestions for affected functionality. This can be done without a new Meteor release. I'll discuss this with my team, as we may be able to implement it easily now that we understand the impact better.

**Comment by nachocodoner** [comment]
I provided the fix and tests on affected functionality on recent changes in this PR, https://github.com/meteor/meteor/pull/13225

I will notify here when published the new version.

**Comment by StorytellerCZ** [comment]
Good to have the tests. With 3.0 I would like to look at if we still need to maintain underscore instead of deprecating it and directing people to the official release.
If yes I have a PR that will update a few more versions up, I will incorporate @nachocodoner fix into that. I think unless we encounter any additional issues we should then deprecate it.
If we still need to make changes to underscore to account for Meteor specific things then going beyond the version I have in my PR will require quiet a bit of work, it might be better then to pull underscore from npm and override the affected functions.

**Comment by nachocodoner** [comment]
Published underscore@1.6.2!

**Comment by evolross** [comment]
Thanks for doing this. Been patiently waiting on meteor `underscore!` for a _long_ time. It's been double published in our builds forever - meteor version, npm version.

Curious about something... you've published `meteor/underscore@1.6.2!` and I see it listed on [atmosphere](https://atmospherejs.com/meteor/underscore) but where are your actual changes for this release in the repo. What branch are they on?

**Comment by nachocodoner** [comment]
@evolross: https://github.com/meteor/meteor/pull/13225

They are here and about to be merged and spread to release 3.0.
